web/modules/custom/rdf_entity/src/Entity/RdfEntitySparqlStorage.php

```diff
@@ -656,34 +659,64 @@ protected function doSave($id, EntityInterface $entity) {
     $properties_list = "<" . implode(">, <", $properties['flat']) . ">";
     foreach ($entity->toArray() as $field_name => $field) {
       foreach ($field as $field_item) {
         foreach ($field_item as $column => $value) {
           if (!isset($properties['by_field'][$field_name][$column])) {
             continue;
           }
-          $pred = '<' . (string) $properties['by_field'][$field_name][$column] . '>';
-          if (!filter_var($value, FILTER_VALIDATE_URL) === FALSE) {
-            $obj = '<' . $value . '>';
+          // Skip the bundle as it is handled separately later.
+          if ($field_name == 'rid') {
+            continue;
+          }
+          // Don't add empty values.
+          if ($value === "") {
+            continue;
           }
+
+          $item = $entity->get($field_name)->first();
+          if (empty($item)) {
+            continue;
+          }
+          $reference = FALSE;
+          // When the field is a entity reference, and it's target implements
+          // RdfEntitySparqlStorage (it's an RDF based entity),
+          // then store it as a resource.
+          if ($item instanceof EntityReferenceItem) {
+            /** @var \Drupal\Core\Entity\Plugin\DataType\EntityReference $entity_property */
+            $entity_property = $item->get('entity');
+            $target = $entity_property->getTarget();
+            /** @var EntityInterface $target_entity */
+            $target_entity = $target->getValue();
+            $target_entity_type = $target_entity->getEntityType();
+            $target_entity_storage_class = $target_entity_type->getStorageClass();
+            if ($target_entity_storage_class == '\Drupal\rdf_entity\Entity\RdfEntitySparqlStorage') {
+              $reference = TRUE;
+            }
+          }
+          // Entity reference to another rdf entity.
+          if ($reference) {
+            $graph->addResource((string) $id, (string) $properties['by_field'][$field_name][$column], $value);
+          }
+          // All other fields get stored as a literal.
           else {
-            // @todo This is most probably prone to Sparql injection..!
-            $obj = '"""' . $value . '"""';
+            // @todo Set language and field data type.
+            $graph->addLiteral((string) $id, (string) $properties['by_field'][$field_name][$column], $value);
           }
-          $insert .= $subj . ' ' . $pred . ' ' . $obj . '  .' . "\n";
         }
       }
     }
-    // Save the bundle.
+    // Save the bundle as rdf:type.
     $rdf_bundle_mapping = $this->mappingHandler->getRdfBundleMappedUri($entity->getEntityType()->getBundleEntityType(), $entity->bundle());
     $rdf_bundle = $rdf_bundle_mapping[$entity->bundle()];
-    $insert .= $subj . ' ' . $this->rdfBundlePredicate . ' <' . $rdf_bundle . '>  .' . "\n";
+    $graph->addResource((string) $id, $this->rdfBundlePredicate, $rdf_bundle);
 
-    $query = <<<QUERY
+    if (!$entity->isNew()) {
+      $query = <<<QUERY
```

Comment:
Since you moved the insert query to a private method, it would be nice to get read of this piece here to another one.